lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -34,420 +25,546 @@
+                          [&](auto type) { return type.getFieldID(index); });
+              return fieldID;
+            };
+            for (auto [index, v] : llvm::enumerate(op.getOperands()))
+              recordValueRefersToFieldRef(res, getFieldId(index), v);
+          })
+          .Case<FConnectLike>([&](FConnectLike connect) {
+            recordDataflow(connect.getDest(), connect.getSrc());
+          })
+          .Default([&](Operation *op) {
+            // All other expressions.
+            if (op->getNumResults() == 1) {
+              auto res = op->getResult(0);
+              for (auto src : op->getOperands())
+                recordDataflow(res, src);
+            }
+          });
     }
   }
 
-  void handleInstanceOp(InstanceOp ins, SmallVector<Value> &worklist) {
-    for (auto port : ins.getResults()) {
-      if (auto type = type_dyn_cast<FIRRTLBaseType>(port.getType())) {
-        worklist.push_back(port);
-        if (!type.isGround())
-          setValRefsTo(port, FieldRef(port, 0));
-      } else if (auto type = type_dyn_cast<PropertyType>(port.getType())) {
-        worklist.push_back(port);
-      }
-    }
+  std::string getName(FieldRef v) { return getFieldName(v).first; };
+
+  unsigned getOrAddNode(Value v) {
+    auto iter = valToFieldRefs.find(v);
+    if (iter == valToFieldRefs.end())
+      return getOrAddNode({v, 0});
+    return getOrAddNode(*iter->second.begin());
   }
 
-  void handlePorts(FieldRef ref, SmallVectorImpl<Value> &children) {
-    if (auto inst = dyn_cast_or_null<InstanceOp>(ref.getDefiningOp())) {
-      auto res = cast<OpResult>(ref.getValue());
-      auto portNum = res.getResultNumber();
-      auto refMod =
-          dyn_cast_or_null<FModuleOp>(*instanceGraph.getReferencedModule(inst));
-      if (!refMod)
-        return;
-      FieldRef modArg(refMod.getArgument(portNum), ref.getFieldID());
-      auto pathIter = portPaths.find(modArg);
-      if (pathIter == portPaths.end())
+  // Get the node id if it exists, else add it to the graph.
+  unsigned getOrAddNode(FieldRef f) {
+    auto iter = nodes.find(f);
+    if (iter != nodes.end())
+      return iter->second;
+    // Add the fieldRef to the graph.
+    auto id = reachingDefGraph.size();
+    // The node id can be used to index into the graph. The entry is a pair,
+    // first element is the corresponding FieldRef, and the second entry is a
+    // list of adjacent nodes.
+    reachingDefGraph.push_back({f, {}});
+    nodes[f] = id;
+    return id;
+  }
+
+  // Construct the reaching def graph, by adding `dst` and `src` as new nodes,
+  // if not already existing. Then add an edge from `src` to `dst`.
+  void addReachingDef(FieldRef dst, FieldRef src) {
+    auto srcNode = getOrAddNode(src);
+    auto dstNode = getOrAddNode(dst);
+    reachingDefGraph[dstNode].second.push_back(srcNode);
+  }
+
+  // Add `srcVal` as a reaching def to `dstVal`.
+  void recordDataflow(Value dstVal, Value srcVal) {
```

Comment:
I think this in generally is not properly handling flip in aggregates.  I am finding that this is not reachable from the `firtool` pipeline, due to lowering of aggregates and how we generate the IR for connects.  But here is an example that is broken:
`./bin/circt-opt --pass-pipeline='builtin.module(firrtl.circuit(firrtl-check-comb-loops))' ./test3.mlir`
```mlir
firrtl.circuit "Test" {
  firrtl.module @Test(){
    %w = firrtl.wire : !firrtl.bundle<a flip: uint<8>>
    %x = firrtl.wire : !firrtl.bundle<a flip: uint<8>>
    // x.a <= w.a
    firrtl.connect %w, %x: !firrtl.bundle<a flip: uint<8>>, !firrtl.bundle<a flip: uint<8>>
    // w.a <= x.a
    %w_a = firrtl.subfield %w[a] : !firrtl.bundle<a flip: uint<8>>
    %x_a = firrtl.subfield %x[a] : !firrtl.bundle<a flip: uint<8>>
    firrtl.strictconnect %w_a, %x_a : !firrtl.uint<8>
  }
}
```

Reply:
Thanks for the test case, this implementation does not handle flips yet.